java/ql/src/experimental/Security/CWE/CWE-706/UserControlledArbitraryRead.ql

```diff
@@ -0,0 +1,221 @@
+/**
+ * @name Disclosure of user-controlled path expression
+ * @description Disclosing content from paths influenced by users can allow an attacker to read arbitrary resources.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision medium
+ * @id java/tainted-file-read
+ * @tags security
+ *       external/cwe/cwe-706
+ */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking2
+import semmle.code.java.security.XSS
+import DataFlow2::PathGraph
+import PathsCommon
+
+/** The class `org.json.JSONObject`. */
+class TypeJsonObject extends Class {
+  TypeJsonObject() { this.hasQualifiedName("org.json", "JSONObject") }
+}
+
+/** The class `org.json.JSONArray`. */
+class TypeJsonArray extends Class {
+  TypeJsonArray() { this.hasQualifiedName("org.json", "JSONArray") }
+}
+
+/** The class `ai.susi.server.ServiceResponse`. */
+class TypeServiceResponse extends Class {
+  TypeServiceResponse() { this.hasQualifiedName("ai.susi.server", "ServiceResponse") }
+}
+
+class ServiceResponseSink extends DataFlow::ExprNode {
+  ServiceResponseSink() {
+    exists(ConstructorCall call |
+      call.getConstructedType() instanceof TypeServiceResponse and
+      this.getExpr() = call.getAnArgument()
+    )
+    or
+    exists(MethodAccess call |
+      call.getType() instanceof TypeServiceResponse and
+      this.getExpr() = call.getAnArgument()
+    )
+  }
+}
+
+predicate deletesFile(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFile and
+    call.getMethod().getName().matches("delete%") and
+    node.getExpr() = call.getQualifier()
+  )
+}
+
+predicate deletesPath(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFiles and
+    call.getMethod().getName().matches("delete%") and
+    node.getExpr() = call.getArgument(0)
+  )
+}
+
+predicate renamesFile(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFile and
+    call.getMethod().getName().matches("renameTo%") and
+    (
+      node.getExpr() = call.getQualifier()
+      or
+      node.getExpr() = call.getArgument(0)
+    )
+  )
+}
+
+predicate renamesPath(DataFlow::ExprNode node) {
+  exists(MethodAccess call |
+    call.getReceiverType() instanceof TypeFiles and
+    call.getMethod().getName().matches("move%") and
+    (
+      node.getExpr() = call.getArgument(0)
+      or
+      node.getExpr() = call.getArgument(1)
+    )
+  )
+}
+
+class SensitiveFileOperationSink extends DataFlow::ExprNode {
```

Comment:
Maybe consider the following as well?
- `Files.setAttributes`
- `Files.setOwner`
- `Files.setPosixFilePermissions`

And same for respective `java.io.File` methods.